www/js/equation.js

```diff
@@ -48,5 +49,132 @@ define([
+		 equations and inputs of existing nodes.
+		 */
+		addQuantity: function(id, subModel){
+
+			var name = subModel.getName(id);
+			array.forEach(subModel.getNodes(), function(node){
+				if(node.equation){
+					try {
+						var expr = Parser.parse(node.equation);
+					}catch(e){
+						/* If an equation fails to parse, then the input
+						 string is stored as the equation for that node.
+						 Thus, if the parse fails, just move on to the 
+						 next node. */
+						return;
+					}
+					var changed = false;
+					array.forEach(expr.variables(), function(variable){
+						if(name == variable){
+							changed = true;
+							expr.substitute(name, id);
+						}
+					});
+					if(changed){
+						node.equation = expr.toString(true);
+						node.inputs = [];
+						var inputs = this.createInputs(expr);
+						array.forEach(inputs, function(input){
+							if(subModel.isNode(input.ID))
+								node.inputs.push(input);
+						});
+					}
+				}
+			},this);
+		},
+
+		// Test if this is a pure sum or product
+		// If so, determine connection labels
+		createInputs: function(parse){
+			var grad;
+			var chooseSign = function(x, a, b, c){
+				return x>0.5?a:(x<-0.5?b:c);
+			};
+			if(this.isSum(parse)){
+				grad = this.gradient(parse, false);
+				return array.map(parse.variables(), function(x){
+					return {ID: x, label: chooseSign(grad[x],"","-","0")};
+				});
+			}else if(this.isProduct(parse)){
+				grad = this.gradient(parse, true);
+				return array.map(parse.variables(), function(x){
+					return {ID: x, label: chooseSign(grad[x],"","/","none")};
+				});
+			}else{
+				// General expression
+				return array.map(parse.variables(), function(x){
+					return {ID: x};
+				});
+			}
+		},
+
+		isSum: function(parse){
+			// Return true if expression is a sum of variables, allowing for minus signs.
+			// Note that a bare variable will also return true.
+			var ops = parse.operators();
+			var allowed = {"+": true, "-": true, "variable": true};
+			for(var op in ops){
+				if(ops[op] > 0 && !allowed[op])
+					return false;
+			}
+			return true;
+		},
+
+		isProduct: function(parse){
+			// Return true if the expression is a product of variables, allowing for division
+			// Note that explicit powers (a^2) are not allowed, which is mathematically incorrect
+			// but we have no mechanism for adding powers on our user interface.  For problems
+			// that are that complicated, the student should be using the full text entry anyway.
+			// Note that a bare variable will also return true.
+			var ops = parse.operators();
+			var allowed = {"*": true, "/": true, "variable": true};
+			for(var op in ops){
+				if(ops[op] > 0 && !allowed[op])
+					return false;
+			}
+			return true;
+		},
+		
+        gradient: function(parse, /*boolean*/ monomial, point){
+			// Find the numerical partial derivatives of the expression at
+			// the given point or at a random point, if the point is not supplied.
```

Comment:
This is correct, we don't have the signs any more, so we can take this function out for now.  @riteshsamala please remove.